bin/grep.py

```diff
@@ -3,22 +3,78 @@
 
 from __future__ import print_function
 
-import argparse
-import collections
-import fileinput
 import os
+import io
 import re
 import sys
+import argparse
+import fileinput
+import fnmatch
+from six import PY3
+
+if PY3:
+    from pathlib import Path
+else:
+    from pathlib2 import Path
+
+_stash = globals()['_stash']
+abbreviate = _stash.libcore.abbreviate
+
+new_file = True
+count = 0        
+ns = None
+fn = None
 
+def open_hook(filename,mode):
+    global fn
+    global new_file
+    global count
+    global ns
+    global abbreviate
```

Comment:
You seem to have a partial wrong understanding on how python `global` works, or more specifically when you'd have to add `global`.

In this case `abbreviate` seems to point to `_stash.libcore.abbreviate`. As abbreviate is assigned to at module-level **and** you do not re-assign it in `open_hook`, no `global` is needed. By default, functions (and similar) can access global variables/functions/... without it being explicitly marked as global using `global`. The exception is if said you would also assign to the variable/function inside of the function, in which case python would treat the variable/function as a local unless explicitly changed using `global`.

So, to recap: you do not need `global abbreviate` as you do not assign to it. However, if you had **instead** written `abbreviate = ...` somewhere in your code, `global abbreviate` would be necessary. This is so that you don't accidentally overwrite a function or global from within a function just because some name happen to be the same.

BTW, I believe the reason variables defined on module-level are global by default is because fundamentally, functions are also objects in python and objects can be used as a function.

Still, as written above you shouldn't use `global` here. For `ns` and `abbreviate` it's unnecessary and it removes a fail safe against accidental changing the variables. For `count` and the other application state variables, a class is better suited.